Slack Integration

chloealee opened the pull request:
#### What's this PR do?
Work done with @sbsmith86 , @katiecrane , @DFurnes , @ngjo , and @lkpttn during HEXUPLE coding! 

When a post is tagged as "Good For Storytelling", Rogue sends a Slack nofitication with the Post's info.! 

#### How should this be reviewed?
:eyes:

#### Any background context you want to provide?
Before we merge, we'll need to find out what channel people want this to be in (#badass-members?) and then re-do the webhook portion of this work. 

#### Relevant tickets
Fixes [this](https://www.pivotaltracker.com/n/projects/2019429/stories/152084069) Pivotal ticket

#### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

chloealee commented on app/Http/Controllers/TagsController.php:
```diff
@@ -5,6 +5,7 @@
 use Rogue\Repositories\PostRepository;
 use Rogue\Http\Transformers\PostTransformer;
 use Rogue\Http\Controllers\Traits\TagsRequests;
+use Rogue\Events\PostTagged;
```
do we need this here?

ngjo commented:
@chloealee can we add a line that we'll be continuing to add other post information to this slack notification (like member info!)

chloealee commented:
@ngjo sure thing! Do you mean in the message that gets posted to Slack? So maybe in the link "Click here for the full post on Rogue" we can change to "Click here for the full post on Rogue (we'll be continuing to add other post info. to this Slack notification (like member info!)" That might be too long though? 
![screen shot 2017-12-13 at 11 20 27 am](https://user-images.githubusercontent.com/9019452/33949511-ccfcae1c-dff7-11e7-8b52-0384cde5a625.png)

lkpttn commented on app/Notifications/SlackTagNotification.php:
```diff
@@ -0,0 +1,71 @@
+                                    'Caption' => str_limit($this->post->caption, 140),
+                                    'Why Participated' => str_limit($this->post->signup->why_participated),
+                                ])
+                                ->color('#4e2b63')
```
We should change this to 
`->color(array_random(['#FCD116', '#23b7fb', '#4e2b63']))`
so we get a different brand color each time!

chloealee replied:
love it! updated! :)

katiecrane commented on app/Notifications/SlackTagNotification.php:
```diff
@@ -0,0 +1,71 @@
+     */
+    public function toSlack($notifiable)
+    {
+        return (new SlackMessage)
```
Should we only do this if the Slack URL env var is set?

chloealee replied:
great catch - updated!

katiecrane commented on app/Notifications/SlackTagNotification.php:
```diff
@@ -0,0 +1,73 @@
+     */
+    public function toSlack($notifiable)
+    {
+        if (isset(SLACK_WEBHOOK_INTEGRATION_URL)) {
```
I thiiiiink you can use `if (config('services.slack.url'))`?

ngjo commented:
@chloealee maybe: Click here for more details in Rogue! We'll be adding member info here soon.

??

DFurnes commented on app/Notifications/SlackTagNotification.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        if (config('services.slack.url')) {
+            return (new SlackMessage)
+                ->from('Tongue Cat', ':tonguecat:')
```
This is a lame suggestion, but would it make sense to have the post by "Rogue" instead of "Tongue Cat"? (Thinking that could help more people on staff associate what Rogue is used for.)

chloealee commented:
@ngjo you're a wordsmith! Changed it and here is what a looks like now (@DFurnes also updated to Rogue instead of from Tongue Cat!):
![screen shot 2017-12-13 at 1 13 11 pm](https://user-images.githubusercontent.com/9019452/33954792-7aa11008-e007-11e7-850c-725902c479c6.png)

chloealee commented:
yay! Ok so I think everything is all set to go besides re-running the webhook to route to the correct room in Slack. @ngjo should we have this go to #badass-members?

katiecrane commented:
This is kind of nitty gritty but to me "we'll be adding member info here soon" sounds like it is being added in Rogue, so I wonder if it would be clearer if that said "We'll be adding member info in Slack soon" 🤔

chloealee commented on .env.example:
```diff
@@ -54,3 +54,5 @@ PHOENIX_API_VERSION=v1
 BLINK_USERNAME=
 BLINK_PASSWORD=
+
+SLACK_WEBHOOK_INTEGRATION_URL=
```
I actually could just update the room in the Slack user interface but use the same URL! Updated this here since this is a new thing we'll need. Should I add this to Heroku now? 

After that's added, this seems like it's good to go! Unless I should test? But that might ruin the surprise/could be the surprise reveal?!

DFurnes replied:
It couldn't hurt to test one more time in `#team-bleed`, but then let's do it!!!

katiecrane replied:
@chloealee I think we only want the url to be set for production so tags from Thor and Staging don't hit `#badass_members` ? Unless we can use a different url that sends to #team-bleed for us to test stuff?

chloealee replied:
@katiecrane ah so smart!! I'll remove those all those except for on Prod after I test on Thor!

DFurnes commented on app/Events/PostTagged.php:
```diff
@@ -0,0 +1,40 @@
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
+
+    public $post;
```
Can we add docblocks for these two properties?

DFurnes commented on app/Events/PostTagged.php:
```diff
@@ -0,0 +1,40 @@
+
+class PostTagged
+{
+    use Dispatchable, InteractsWithSockets, SerializesModels;
```
Since this event is internal to the application (it's just triggering a listener class), I think we can kill `InteractsWithSockets` and the `broadcastOn` methods here. 📻

DFurnes commented on app/Notifications/SlackTagNotification.php:
```diff
@@ -0,0 +1,73 @@
+     */
+    public function toSlack($notifiable)
+    {
+        if (config('services.slack.url')) {
```
Nitpicky, but it could be cleaner to reformat this as a [guard clause](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html):

```php
if (! config('services.slack.url')) {
    return;
}

return (new SlackMessage)
    ->from('Rogue', ':tonguecat:')
    // ...
```
